emu/src/single_data_transfer.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{arm7tdmi::Arm7tdmi, bitwise::Bits, io_device::IoDevice, opcode::ArmModeOpcode};
+
+/// Possible opeartion on transfer data.
+enum SingleDataTransfer {
+    /// Load from memory into a register.
+    Ldr,
+
+    /// Store from a register into memory.
+    Str,
+    Pld,
+}
+
+impl From<u32> for SingleDataTransfer {
+    fn from(op_code: u32) -> Self {
+        let must_for_pld = op_code.are_bits_on(28..=31);
+        if op_code.get_bit(20) {
+            if must_for_pld {
+                Self::Pld
+            } else {
+                Self::Ldr
+            }
+        } else {
+            Self::Str
+        }
+    }
+}
+
+/// There two different kind of write or read for memory.
+#[derive(Default)]
+enum ReadWriteKind {
+    /// Word is a u32 value for ARM mode and u16 for Thumb mode.
+    #[default]
+    Word,
+
+    /// Byte is a u8 value.
+    Byte,
+}
+
+impl From<bool> for ReadWriteKind {
+    fn from(value: bool) -> Self {
+        if value {
+            Self::Byte
+        } else {
+            Self::Word
+        }
+    }
+}
+
+impl From<u32> for ReadWriteKind {
+    fn from(op_code: u32) -> Self {
+        op_code.get_bit(22).into()
+    }
+}
+
+impl Arm7tdmi {
+    pub(crate) fn single_data_transfer(&mut self, op_code: ArmModeOpcode) {
+        let immediate = op_code.get_bit(25);
+        let up_down = op_code.get_bit(23);
+        let byte_or_word: ReadWriteKind = op_code.raw.into();
+
+        // bits [19-16] - Base register
+        let rn = op_code.get_bits(16..=19);
+
+        // 0xF is register of PC
+        let address = if rn == 0xF {
+            let pc: u32 = self.registers.program_counter().try_into().unwrap();
+            pc + 8_u32
+        } else {
+            self.registers.register_at(rn.try_into().unwrap())
+        };
+
+        // bits [15-12] - Source/Destination Register
+        let rd = op_code.get_bits(12..=15);
+
+        let offset = if immediate {
+            todo!()
+        } else {
+            op_code.get_bits(0..=11)
+        };
+
+        let load_store: SingleDataTransfer = op_code.raw.into();
```

Comment:
Personally, I don't like the idea of doing `ArmModOpcode` -> `u32` -> `SingleDataTransfer`. Couldn't we just implement `From<ArmModeOpcode` for `SingleDataTransfer` instead of manually going through `u32`?

The same also applies to `ReadWriteKind`.

What do you think?

Reply:
We can change it :)